src/api/udf/RoundTimeToUnit.h

```diff
@@ -0,0 +1,112 @@
+      [unit, beginTime](const std::optional<InputType>& origin) -> std::optional<int> {
+        if (N_UNLIKELY(origin == std::nullopt)) {
+          return std::nullopt;
+        }
+
+        static constexpr int32_t HOUR_CASE = 1;
+        static constexpr int32_t DAY_CASE = 2;
+        static constexpr int32_t WEEK_CASE = 3;
+        static constexpr int32_t MONTH_CASE = 4;
+        static constexpr int32_t QUARTER_CASE = 5;
+        static constexpr int32_t YEAR_CASE = 6;
+
+        // need to subtract begintime
+
+        InputType timeSecs = origin.value();
+        RoundedTimeType res = std::numeric_limits<RoundedTimeType>::lowest();
+        std::time_t timePoint = (std::time_t) timeSecs;
+        switch (unit) {
+            case HOUR_CASE: {
+              std::time_t roundedPoint = nebula::common::Evidence::hour(timePoint); 
+              // LOG(INFO) << "FORMATTED ROUNDED POINT: " << nebula::common::Evidence::fmt_normal(roundedPoint);
+              res = (RoundedTimeType) roundedPoint;
+              break;
+            }
+            case DAY_CASE: {
+              std::time_t roundedPoint = nebula::common::Evidence::date(timePoint); 
+              // LOG(INFO) << "FORMATTED ROUNDED POINT: " << nebula::common::Evidence::fmt_normal(roundedPoint);
+              res = (RoundedTimeType) roundedPoint;
+              break;
+            }
+            case WEEK_CASE: { 
+              std::time_t roundedPoint = nebula::common::Evidence::week(timePoint); 
+              // LOG(INFO) << "FORMATTED ROUNDED POINT: " << nebula::common::Evidence::fmt_normal(roundedPoint);
+              res = (RoundedTimeType) roundedPoint;
+              break;
+            }
+            case MONTH_CASE: { 
+                std::time_t roundedPoint = nebula::common::Evidence::month(timePoint); 
+                // LOG(INFO) << "FORMATTED ROUNDED POINT: " << nebula::common::Evidence::fmt_normal(roundedPoint);
+                res = (RoundedTimeType) roundedPoint;
+                break;
+            }
+            case QUARTER_CASE: {
+                break;
+            }
+            case YEAR_CASE: { 
+                std::time_t roundedPoint = nebula::common::Evidence::year(timePoint); 
+                // LOG(INFO) << "FORMATTED ROUNDED POINT: " << nebula::common::Evidence::fmt_normal(roundedPoint);
+                res = (RoundedTimeType) roundedPoint;
+                break;
+            }
+        }
+
+        N_ENSURE_NE(res, std::numeric_limits<RoundedTimeType>::lowest(), "res should be defined");
+        
+        // LOG(INFO) << "timeSecs: " << nebula::common::Evidence::fmt_normal(timeSecs);
+        // LOG(INFO) << "unit: " << unit;
+        // LOG(INFO) << "res: " << nebula::common::Evidence::fmt_normal(res);
+        // LOG(INFO) << "(res - beginTime): " << nebula::common::Evidence::fmt_normal(res - beginTime);
+
+        RoundedTimeType minValue = (RoundedTimeType)(0);
+
+        return std::max(minValue, res - beginTime);
```

Comment:
can you add one line comment to make sure this is the contract that was agreed on - seems very important to make sure the value returned is positive.